aggregator/aggregator.go

```diff
@@ -43,275 +40,238 @@ func NewAggregator(
+	}
+	a.compareInputHashes(inputProver, resGetProof)
+
+	log.Infof("sending verified proof to the ethereum smart contract, batchNumber", batchToVerify.BatchNumber)
+	err = a.EthTxManager.VerifyBatch(batchToVerify.BatchNumber, resGetProof)
+	if err != nil {
+		log.Warnf("failed to send request to consolidate batch to ethereum, batch number: %d, err: %v",
+			batchToVerify.BatchNumber, err)
+		return
+	}
+	a.batchesSent[batchToVerify.BatchNumber] = true
+	log.Infof("proof for the batch was send, batchNumber: %d", batchToVerify.BatchNumber)
+}
+
+func (a *Aggregator) isSynced(ctx context.Context) bool {
+	lastVerifiedBatch, err := a.State.GetLastVerifiedBatch(ctx, nil)
+	if err != nil && err != state.ErrNotFound {
+		log.Warnf("failed to get last consolidated batch, err: %v", err)
+		return false
+	}
+	if lastVerifiedBatch != nil {
+		a.lastVerifiedBatchNum = lastVerifiedBatch.BatchNumber
+	}
+	lastVerifiedEthBatchNum, err := a.Ethman.GetLatestVerifiedBatchNum()
+	if err != nil {
+		log.Warnf("failed to get last eth batch, err: %v", err)
+		return false
+	}
+	if a.lastVerifiedBatchNum < lastVerifiedEthBatchNum {
+		log.Infof("waiting for the state to be synced, lastVerifiedBatchNum: %d, lastVerifiedEthBatchNum: %d",
+			a.lastVerifiedBatchNum, lastVerifiedEthBatchNum)
+		return false
+	}
+	return true
+}
+
+func (a *Aggregator) getBatchToVerify(ctx context.Context) (*state.Batch, error) {
+	delete(a.batchesSent, a.lastVerifiedBatchNum)
+
+	batchToVerify, err := a.State.GetVirtualBatchByNumber(ctx, a.lastVerifiedBatchNum+1, nil)
+
+	if err != nil {
+		if errors.Is(err, state.ErrNotFound) {
+			log.Infof("there are no batches to consolidate")
+			return nil, err
 		}
+		log.Warnf("failed to get batch to consolidate, err: %v", err)
+		return nil, err
+	}
+
+	if a.batchesSent[batchToVerify.BatchNumber] {
+		log.Infof("batch with number %d was already sent, but not yet consolidated by synchronizer",
+			batchToVerify.BatchNumber)
+		return nil, nil
+	}
+
+	return batchToVerify, nil
+}
+
+func (a *Aggregator) buildInputProver(ctx context.Context, batchToVerify *state.Batch) (*pb.InputProver, error) {
+	oldStateRoot, err := a.State.GetStateRootByBatchNumber(ctx, a.lastVerifiedBatchNum, nil)
+	if err != nil && err != state.ErrNotFound {
+		return nil, fmt.Errorf("failed to get current state root, err: %v", err)
+	}
+
+	newStateRoot, err := a.State.GetStateRootByBatchNumber(ctx, batchToVerify.BatchNumber, nil)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get state root to consolidate, err: %v", err)
+	}
+
+	rawTxs, err := state.EncodeTransactions(batchToVerify.Transactions)
+	if err != nil {
+		return nil, fmt.Errorf("failed to encode transactions, err: %v", err)
+	}
+	globalExitRoot := batchToVerify.GlobalExitRoot
+
+	oldLocalExitRoot, err := a.State.GetLocalExitRootByBatchNumber(ctx, a.lastVerifiedBatchNum, nil)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get local exit root for batch %d, err: %v", a.lastVerifiedBatchNum, err)
+	}
```

Comment:
Get full batch in previous step so you can do something like:

```suggestion
	oldLocalExitRoot := previousBatch.LocalExitRoot
```